HashRegistrar.sol

```diff
@@ -47,34 +47,52 @@ contract Deed {
     address constant burn = 0xdead;
     uint public creationDate;
     address public owner;
-    
+    event OwnerChanged(address newOwner);
+    event deedClosed();
+    bool active;
+
     modifier onlyRegistrar {
         if (msg.sender != address(registrar)) throw;
         _
     }
+
+    modifier onlyActive {
+        if (!active) throw;
+        _
+    }
     
     function Deed() {
         registrar = Registrar(msg.sender);
         creationDate = now;
+        active = true;
     }
         
     function setOwner(address newOwner) {
         if ((owner > 0 && msg.sender != owner )
             || (owner == 0 && msg.sender != address(registrar))) throw;
         owner = newOwner;
+        OwnerChanged(newOwner);
     }
     
-    function setBalance(uint newValue) onlyRegistrar {
+    function setBalance(uint newValue) onlyRegistrar onlyActive {
         // Check if it has enough balance to set the value
         if (this.balance < newValue) throw;
         // Send the difference to the owner
         if (!owner.send(this.balance - newValue)) throw;
     }
 
-    function closeDeed(uint refundRatio) onlyRegistrar {
-        owner.send((refundRatio * this.balance)/1000);
-        selfdestruct(burn);
+    function closeDeed(uint refundRatio) onlyRegistrar onlyActive {
+        active = false;            
+        burn.send(((1000 - refundRatio) * this.balance)/1000);
+        deedClosed();
+        destroyDeed();
     }    
+
+    function destroyDeed() {
```

Comment:
I wonder if changing to a pure withdrawal model wouldn't make this simpler: have `setBalance` only update a balance variable, and have a `withdraw` method, only callable by the owner, that sends them any extra and self-destructs iff the recorded balance is zero (on the assumption that deposits are always nonzero).

That would restrict sends to just that method, and remove the need to reason about who's calling things like setBalance and whether they have incentive to make life difficult; on the other hand, it would require callers to interact with their deeds directly on a day-to-day basis.

What do you think?